zerver/views/message_send.py

```diff
@@ -339,3 +347,65 @@ def render_message_backend(
 
     rendering_result = render_markdown(message, content, realm=user_profile.realm)
     return json_success(request, data={"rendered": rendering_result.rendered_content})
+
+
+REPORT_MESSAGE_REASONS = ["spam", "harassment", "inappropriate", "norms", "other"]
+
+
+@has_request_variables
+def report_message_backend(
+    request: HttpRequest,
+    user_profile: UserProfile,
+    message_id: int = REQ(converter=to_non_negative_int, path_only=True),
+    reason: str = REQ(default="other", str_validator=check_string_in(REPORT_MESSAGE_REASONS)),
+    explanation: Optional[str] = REQ(default=None),
+    # quoting is provided by client-side javascript, i.e. no server-side library
+    quoted_message: Optional[str] = REQ(default=None),
+) -> HttpResponse:
+    request_notes = RequestNotes.get_notes(request)
+    assert request_notes.log_data is not None
+    origmsg, ignored_user_message = access_message(user_profile, message_id, lock_message=False)
+    quoted_message = "" if quoted_message is None else quoted_message.strip()
+    explanation = "" if explanation is None else explanation.strip()
+
+    # todo: link/tooltip the reason to docs explaining what this means
+    rptmsg = f"{_('Reporting reason')}: **`"
+    if reason == "spam":
+        rptmsg += f"{_('Spam')}"
+    elif reason == "harassment":
+        rptmsg += f"{_('Harassment')}"
+    elif reason == "inappropriate":
+        rptmsg += f"{_('Inappropriate content')}"
+    elif reason == "norms":
+        rptmsg += f"{_('Violates community norms')}"
+    else:
+        rptmsg += f"{_('Other reason')}"
+        if explanation == "":
+            raise JsonableError(_("For reason=other, an explanation must be provided."))
+    rptmsg += f"`** by: @_**{user_profile.full_name}|{user_profile.id}** "
+    if explanation != "":
+        rptmsg += f" {_('who included these notes')}:\n```\n{explanation}\n```\n"
```

Comment:
This type of string construction, with a sentence split into multiple translated strings, unfortunately doesn't work with internationalization, because other languages have different word orders.

See https://zulip.readthedocs.io/en/latest/translating/internationalization.html for some background.

Probably we should discuss the formatting in the Zulip thread for the issue, since we might as well polish the sentences at the same time.

Reply:
fyi CZO: https://chat.zulip.org/#narrow/stream/101-design/topic/report.20message.20feature.20.28.2320047.29
"just use a single string template" - I'm on it.

Reply:
<img width="895" alt="image" src="https://user-images.githubusercontent.com/140002/173375186-65b7b4f4-c4cf-4154-ad97-209d05603290.png">

python template: ```f"""@_**{username}|{userid}** reported a message sent by @_**{sendername}|{senderid}** {_('as containing')} **`{reason_str}`**{explanation_str}
{_('Original message')}:
{quoted_message}"""```